SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverRPC.cpp

```diff
@@ -0,0 +1,471 @@
+
+#include "Interop/RPCs/RPCQueues.h"
+
+using namespace SpatialGDK;
+
+void FRPCMetaData::ComputeSpanId(FName Name, SpatialGDK::SpatialEventTracer& Tracer, SpatialGDK::EntityComponentId EntityComponent,
+								 uint64 RPCId)
+{
+	TArray<FSpatialGDKSpanId> ComponentUpdateSpans = Tracer.GetAndConsumeSpansForComponent(EntityComponent);
+
+	SpanId = Tracer.TraceEvent(
+		FSpatialTraceEventBuilder::CreateReceiveRPC(EventTraceUniqueId::GenerateForNamedRPC(EntityComponent.EntityId, Name, RPCId)),
+		/* Causes */ reinterpret_cast<const Trace_SpanIdType*>(ComponentUpdateSpans.GetData()),
+		/* NumCauses */ ComponentUpdateSpans.Num());
+}
+
+void FRPCPayload::ReadFromSchema(Schema_Object* RPCObject)
+{
+	Offset = Schema_GetUint32(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_OFFSET_ID);
+	Index = Schema_GetUint32(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_INDEX_ID);
+	PayloadData = GetBytesFromSchema(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_PAYLOAD_ID);
+}
+
+void FRPCPayload::WriteToSchema(Schema_Object* RPCObject) const
+{
+	Schema_AddUint32(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_OFFSET_ID, Offset);
+	Schema_AddUint32(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_INDEX_ID, Index);
+	AddBytesToSchema(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_PAYLOAD_ID, PayloadData.GetData(), PayloadData.Num());
+}
+
+USpatialNetDriverRPC::USpatialNetDriverRPC() {}
+
+void USpatialNetDriverRPC::OnRPCSent(SpatialGDK::SpatialEventTracer* EventTracer, TArray<UpdateToSend>& OutUpdates, FName Name,
+									 Worker_EntityId EntityId, Worker_ComponentId ComponentId, uint64 RPCId,
+									 const FSpatialGDKSpanId& SpanId)
+{
+	EventTraceUniqueId LinearTraceId = EventTraceUniqueId::GenerateForNamedRPC(EntityId, Name, RPCId);
+	FSpatialGDKSpanId NewSpanId = EventTracer->TraceEvent(FSpatialTraceEventBuilder::CreateSendRPC(LinearTraceId),
+														  /* Causes */ SpanId.GetConstId(), /* NumCauses */ 1);
+
+	if (OutUpdates.Num() == 0 || OutUpdates.Last().EntityId != EntityId || OutUpdates.Last().Update.component_id != ComponentId)
+	{
+		OutUpdates.AddDefaulted();
+		OutUpdates.Last().EntityId = EntityId;
+		OutUpdates.Last().Update.component_id = ComponentId;
+	}
+	OutUpdates.Last().Spans.Add(NewSpanId);
+}
+
+void USpatialNetDriverRPC::OnDataWritten(TArray<FWorkerComponentData>& OutArray, Worker_EntityId EntityId, Worker_ComponentId ComponentId,
+										 Schema_ComponentData* InData)
+{
+	if (ensure(InData != nullptr))
+	{
+		FWorkerComponentData Data;
+		Data.component_id = ComponentId;
+		Data.schema_type = InData;
+		OutArray.Add(Data);
+	}
+}
+
+void USpatialNetDriverRPC::OnUpdateWritten(TArray<UpdateToSend>& OutUpdates, Worker_EntityId EntityId, Worker_ComponentId ComponentId,
+										   Schema_ComponentUpdate* InUpdate)
+{
+	if (ensure(InUpdate != nullptr))
+	{
+		if (OutUpdates.Num() == 0 || OutUpdates.Last().EntityId != EntityId || OutUpdates.Last().Update.component_id != ComponentId)
```

Comment:
nit - you could restructure this code, using AddDefaulted_GetRef, to minimum use of Last()